File: src/python/bubble.py
import os
import time
import struct
import sys

def main():
    try:
        N = int(os.environ.get("SORT_SIZE", 10000))
    except:
        N = 10000

    filename = "data.bin"
    if not os.path.exists(filename):
        print(f"Error: '{filename}' not found. Run 'python3 tests/gen_data.py' first.", file=sys.stderr)
        sys.exit(1)

    with open(filename, "rb") as f:
        data = f.read(N * 8)
        
    # unpack requires exact bytes for the format string
    # Calculate how many doubles we actually read
    actual_n = len(data) // 8
    if actual_n < N:
        print(f"Warning: requested {N}, but file only has {actual_n}", file=sys.stderr)
        # Pad with zeros or adjust N? 
        # For Bubble Sort consistency, let's just use what we have or fill.
        # Here we adjust list to be N size, filling rest with 0.0
        floats = list(struct.unpack(f'<{actual_n}d', data))
        arr = floats + [0.0] * (N - actual_n)
    else:
        arr = list(struct.unpack(f'<{N}d', data))

    start = time.perf_counter()
    
    # Pure Python Bubble Sort
    for i in range(N - 1):
        for j in range(N - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

    end = time.perf_counter()

    print(f"Sort({N}): ", end="")
    p_limit = 5 if N >= 5 else N
    for i in range(p_limit): print(f"{arr[i]:.4f} ", end="")
    print("... ", end="")
    if N > 5:
        for i in range(N-5, N): print(f"{arr[i]:.4f} ", end="")
    print()
    
    print(f"Time: {(end - start) * 1000:.3f} ms")

if __name__ == "__main__":
    main()

File: src/c/bubble.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdint.h>

int main() {
    int N = 10000;
    char *n_env = getenv("SORT_SIZE");
    if (n_env) N = atoi(n_env);

    double *arr = (double*)malloc(N * sizeof(double));
    if (!arr) { perror("malloc"); return 1; }

    // Read from binary file
    FILE *f = fopen("data.bin", "rb");
    if (!f) {
        fprintf(stderr, "Error: 'data.bin' not found. Run 'python3 tests/gen_data.py' first.\n");
        free(arr);
        return 1;
    }
    size_t read_count = fread(arr, sizeof(double), N, f);
    fclose(f);

    if (read_count < N) {
        fprintf(stderr, "Warning: File contained fewer elements (%lu) than requested (%d).\n", read_count, N);
        // We continue with what we read, or update N
        N = read_count;
    }

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    // Bubble Sort
    for (int i = 0; i < N - 1; i++) {
        for (int j = 0; j < N - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                double temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &end);
    double time_ms = (end.tv_sec - start.tv_sec) * 1000.0 + (end.tv_nsec - start.tv_nsec) / 1000000.0;

    printf("Sort(%d): ", N);
    int print_limit = (N < 5) ? N : 5;
    for(int i=0; i<print_limit; i++) printf("%.4f ", arr[i]);
    printf("... ");
    if (N > 5) {
        for(int i=N-5; i<N; i++) printf("%.4f ", arr[i]);
    }
    printf("\n");
    
    printf("Time: %.3f ms\n", time_ms);
    free(arr);
    return 0;
}
